Kernel/FileSystem/ProcFS.cpp

```diff
@@ -239,4 +232,363 @@ InodeMetadata ProcFSLinkInode::metadata() const
+        return KResult(EIO);
+    }
+
+    MutexLocker locker(m_refresh_lock);
+
+    auto& typed_cached_data = static_cast<ProcFSInodeData&>(*description->data());
+    auto& data_buffer = typed_cached_data.buffer;
+
+    if (!data_buffer || (size_t)offset >= data_buffer->size())
+        return 0;
+
+    ssize_t nread = min(static_cast<off_t>(data_buffer->size() - offset), static_cast<off_t>(count));
+    if (!buffer.write(data_buffer->data() + offset, nread))
+        return KResult(EFAULT);
+
+    return nread;
+}
+RefPtr<Inode> ProcFSProcessPropertyInode::lookup(StringView)
+{
+    VERIFY_NOT_REACHED();
+}
+
+static KResult build_from_cached_data(KBufferBuilder& builder, ProcFSInodeData& cached_data)
+{
+    cached_data.buffer = builder.build();
+    if (!cached_data.buffer)
+        return ENOMEM;
+    return KSuccess;
+}
+
+KResult ProcFSProcessPropertyInode::try_to_acquire_data(Process& process, KBufferBuilder& builder) const
+{
+    // FIXME: Verify process is already ref-counted
+    if (m_parent_sub_directory_type == SegmentedProcFSIndex::ProcessSubDirectory::FileDescriptions) {
+        if (auto result = process.acquire_file_description_link(m_possible_data.property_index, builder); result.is_error())
+            return result.error();
+        return KSuccess;
+    }
+    if (m_parent_sub_directory_type == SegmentedProcFSIndex::ProcessSubDirectory::Stacks) {
+        if (auto result = process.acquire_thread_stack(m_possible_data.property_index, builder); result.is_error())
+            return result.error();
+        return KSuccess;
+    }
+
+    VERIFY(m_parent_sub_directory_type == SegmentedProcFSIndex::ProcessSubDirectory::Reserved);
+    switch (m_possible_data.property_type) {
+    case SegmentedProcFSIndex::MainProcessProperty::Unveil:
+        return process.acquire_fds_stats(builder) ? KSuccess : KResult(EFAULT);
+    case SegmentedProcFSIndex::MainProcessProperty::Pledge:
+        return process.acquire_pledge_stats(builder) ? KSuccess : KResult(EFAULT);
+    case SegmentedProcFSIndex::MainProcessProperty::FileDescriptions:
+        return process.acquire_fds_stats(builder) ? KSuccess : KResult(EFAULT);
+    case SegmentedProcFSIndex::MainProcessProperty::BinaryLink:
+        return process.acquire_binary_link(builder) ? KSuccess : KResult(EFAULT);
+    case SegmentedProcFSIndex::MainProcessProperty::CurrentWorkDirectoryLink:
+        return process.acquire_current_work_directory_link(builder) ? KSuccess : KResult(EFAULT);
+    case SegmentedProcFSIndex::MainProcessProperty::PerformanceEvents:
+        return process.acquire_perf_events(builder) ? KSuccess : KResult(EFAULT);
+    case SegmentedProcFSIndex::MainProcessProperty::VirtualMemoryStats:
+        return process.virtual_memory_stats(builder) ? KSuccess : KResult(EFAULT);
+    case SegmentedProcFSIndex::MainProcessProperty::RootLink:
+        return process.acquire_root_link(builder) ? KSuccess : KResult(EFAULT);
```

Comment:
Why are these using `EFAULT` as the default error?
Let's make them return `KResult` directly instead so we don't need the `?:` dance here.
Also, the word `acquire` sounds like this is lock-related. Maybe we can use a prefix like `procfs_get` instead of `acquire` for these names?